In [8]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)



In [9]:
df_mat=pd.read_csv('Dataset/student-mat.csv',sep=';')


In [10]:
df_mat

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

In [11]:
df_mat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [12]:
for i in df_mat.columns:
    print(df_mat[i].value_counts())

GP    349
MS     46
Name: school, dtype: int64
F    208
M    187
Name: sex, dtype: int64
16    104
17     98
15     82
18     82
19     24
20      3
21      1
22      1
Name: age, dtype: int64
U    307
R     88
Name: address, dtype: int64
GT3    281
LE3    114
Name: famsize, dtype: int64
T    354
A     41
Name: Pstatus, dtype: int64
4    131
2    103
3     99
1     59
0      3
Name: Medu, dtype: int64
2    115
3    100
4     96
1     82
0      2
Name: Fedu, dtype: int64
other       141
services    103
at_home      59
teacher      58
health       34
Name: Mjob, dtype: int64
other       217
services    111
teacher      29
at_home      20
health       18
Name: Fjob, dtype: int64
course        145
home          109
reputation    105
other          36
Name: reason, dtype: int64
mother    273
father     90
other      32
Name: guardian, dtype: int64
1    257
2    107
3     23
4      8
Name: traveltime, dtype: int64
2    198
1    105
3     65
4     27
Name: studytime, dtype: int64
0    312
1  

# Data is imbalanced, We will use Tree based models

# Removing Outliers,if any

In [13]:
result = df_mat.select_dtypes(include='number')

In [14]:
for i in result.columns:
    percentile25 = df_mat[i].quantile(0.25)
    percentile75 = df_mat[i].quantile(0.75)
    
    iqr = percentile75-percentile25
    
    upper_limit = percentile75 + 1.5 * iqr
    lower_limit = percentile25 - 1.5 * iqr
    
    df_mat[df_mat[i] > upper_limit]
    df_mat[df_mat[i] < lower_limit]
    
    df = df_mat[df_mat[i] < upper_limit ]
    df = df_mat[df_mat[i] > lower_limit ]
    


In [15]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

In [16]:
df_mat.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

In [17]:
col=['schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic']
dic={'no':0,'yes':1}

for i in col:
    
    df_mat[i]=df_mat[i].map(dic)

In [18]:
df.corr()['G3'].sort_values()

failures     -0.360415
age          -0.161579
goout        -0.132791
traveltime   -0.117142
health       -0.061335
Dalc         -0.054660
Walc         -0.051939
freetime      0.011307
absences      0.034247
famrel        0.051363
studytime     0.097820
Fedu          0.152457
Medu          0.217147
G1            0.801468
G2            0.904868
G3            1.000000
Name: G3, dtype: float64

* schoolsup    -0.082788
* health       -0.061335
* Dalc         -0.054660
* Walc         -0.051939
* famsup       -0.039157
* freetime      0.011307

**These features are having very less correlation, Lets drop them**

In [19]:
df.drop(columns=['schoolsup','health','Dalc','Walc','famsup','freetime'],inplace=True)

# On client Demand

In [20]:
df.drop(columns=['nursery','romantic'],inplace=True)

In [21]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... paid activities  higher  internet  famrel goout absences  G1  G2  G3  
0    ...   no         no     yes        no       4     4        6   5   6   6  
1    ...   no         no     yes       yes       5     3        4   5   5   6  
2    ...  yes         no     yes       yes       4     2       10   7   8  10  
3    ...  yes        yes     yes       yes       3     2        2  15  14  15  
4    ...  yes         no     yes        no       4     2        4   6  10  10  
..   ...  ...        ...     ...       ...     ...   ...      ...  ..  ..  ..  
390  ...  yes         no     yes        no       5     4       11   9   9   9  
391  ...   no         no     yes       yes       2     5        3  14  16  16  
392  ...   no         no     yes        no       5     3        3  10   8   7  
393  ...   no         no     yes       yes       4     1        0  11  12  10  
394  ...   no         no     yes       yes       3     3        5   8   9   9  

[395 rows x 25 columns]

1. **See value counts of F job and M job ,'other' is almost equal or greater than rest of value. and here other is generic value. For betterment lets drop these columns**

2. **We should also drop school as out of 2 school, 1 has 70% more enrollement**


In [22]:
df=df.drop(columns=['Mjob','Fjob','school'])

In [23]:
df.reason.value_counts()

course        145
home          109
reputation    105
other          36
Name: reason, dtype: int64

In [24]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
df.address=le.fit_transform(df.address)
df.sex=le.fit_transform(df.sex)
df.famsize=le.fit_transform(df.famsize)
df.Pstatus=le.fit_transform(df.Pstatus)
df.paid=le.fit_transform(df.paid)
df.activities=le.fit_transform(df.activities)
df.higher=le.fit_transform(df.higher)
df.internet=le.fit_transform(df.internet)


In [25]:
for i in df_mat.columns:
    print(df_mat[i].value_counts())

GP    349
MS     46
Name: school, dtype: int64
F    208
M    187
Name: sex, dtype: int64
16    104
17     98
15     82
18     82
19     24
20      3
21      1
22      1
Name: age, dtype: int64
U    307
R     88
Name: address, dtype: int64
GT3    281
LE3    114
Name: famsize, dtype: int64
T    354
A     41
Name: Pstatus, dtype: int64
4    131
2    103
3     99
1     59
0      3
Name: Medu, dtype: int64
2    115
3    100
4     96
1     82
0      2
Name: Fedu, dtype: int64
other       141
services    103
at_home      59
teacher      58
health       34
Name: Mjob, dtype: int64
other       217
services    111
teacher      29
at_home      20
health       18
Name: Fjob, dtype: int64
course        145
home          109
reputation    105
other          36
Name: reason, dtype: int64
mother    273
father     90
other      32
Name: guardian, dtype: int64
1    257
2    107
3     23
4      8
Name: traveltime, dtype: int64
2    198
1    105
3     65
4     27
Name: studytime, dtype: int64
0    312
1  

In [26]:
df.guardian=le.fit_transform(df.guardian)

In [27]:
df.corr()['G3'].sort_values()

failures     -0.360415
age          -0.161579
goout        -0.132791
traveltime   -0.117142
guardian     -0.070109
Pstatus      -0.058009
activities    0.016100
absences      0.034247
famrel        0.051363
famsize       0.081407
studytime     0.097820
internet      0.098483
paid          0.101996
sex           0.103456
address       0.105756
Fedu          0.152457
higher        0.182465
Medu          0.217147
G1            0.801468
G2            0.904868
G3            1.000000
Name: G3, dtype: float64

* Correlation

In [28]:
df.drop(columns=['studytime','famsize','famrel','absences','activities','Pstatus','guardian'],inplace=True)

In [29]:
df

sex  age  address  Medu  Fedu  reason  traveltime  failures  paid  \
0      0   18        1     4     4  course           2         0     0   
1      0   17        1     1     1  course           1         0     0   
2      0   15        1     1     1   other           1         3     1   
3      0   15        1     4     2    home           1         0     1   
4      0   16        1     3     3    home           1         0     1   
..   ...  ...      ...   ...   ...     ...         ...       ...   ...   
390    1   20        1     2     2  course           1         2     1   
391    1   17        1     3     1  course           2         0     0   
392    1   21        0     1     1  course           1         3     0   
393    1   18        0     3     2  course           3         0     0   
394    1   19        1     1     1  course           1         0     0   

     higher  internet  goout  G1  G2  G3  
0         1         0      4   5   6   6  
1         1         1      3   5   5   6  
2         1         1      2   7   8  10  
3         1         1      2  15  14  15  
4         1         0      2   6  10  10  
..      ...       ...    ...  ..  ..  ..  
390       1         0      4   9   9   9  
391       1         1      5  14  16  16  
392       1         0      3  10   8   7  
393       1         1      1  11  12  10  
394       1         1      3   8   9   9  

[395 rows x 15 columns]

In [30]:
## one hot encoding the 'Location' column
dummies = pd.get_dummies(df['reason'])
df = pd.concat([df,dummies], axis='columns')
df.drop('reason', axis=1, inplace=True)

In [31]:
df.corr()['G3'].sort_values()

failures     -0.360415
age          -0.161579
goout        -0.132791
traveltime   -0.117142
course       -0.098950
home         -0.021359
other         0.052008
reputation    0.095692
internet      0.098483
paid          0.101996
sex           0.103456
address       0.105756
Fedu          0.152457
higher        0.182465
Medu          0.217147
G1            0.801468
G2            0.904868
G3            1.000000
Name: G3, dtype: float64

In [32]:
df.drop(columns=['course','home','other','reputation'],inplace=True)

In [33]:
from sklearn.model_selection import train_test_split
X=df.drop('G3',axis=1)
Y=df.G3

x_train, x_test, y_train, y_test = train_test_split(X,Y,test_size = .25, random_state = 111)

In [34]:
X.columns

Index(['sex', 'age', 'address', 'Medu', 'Fedu', 'traveltime', 'failures',
       'paid', 'higher', 'internet', 'goout', 'G1', 'G2'],
      dtype='object')

In [35]:
# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
RF=RandomForestRegressor(max_depth=100)
model=RF.fit(x_train,y_train)

In [36]:
import pickle
import pickle
pickle.dump(model, open('model.sav', 'wb'))

In [37]:
y_pred=model.predict(x_test)

In [38]:
from sklearn.metrics import explained_variance_score

explained_variance_score( y_pred,y_test)

0.8207770200569151

In [39]:
def predict_price(sex, age, address, Medu, Fedu, traveltime, failures,
       paid, higher, internet, goout, G1, G2):
   
    x = np.zeros(len(X.columns))
    x[0] = sex 
    x[1] = age
    x[2] = address 
    x[3] = Medu 
    x[4] = Fedu
    x[5] = traveltime 
    x[6] = failures
    x[7] =paid
    x[8] = higher 
    x[9] = internet
    x[10] =  goout
    x[11] =  G1
    x[12] =  G2
    
    
    
    return model.predict([x])[0]

In [40]:
predict_price(1,17, 1, 1,1,1, 1,
       0,1, 1,4, 4,1)

C:\Users\WARRIOR\anaconda3\lib\site-packages\sklearn\base.py:445: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


0.92